File: Excho.Data.CSharp/Repository.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Excho.Data.CSharp
{
    public interface IRepository
    {
        IEnumerable All { get; }
        void Commit(object data);
        void Delete(object data);
    }
    public interface IRepository<out T> : IRepository
    {
        IEnumerable<T> All { get; }
        void Commit(T data);
        void Delete(T data);
    }
}

File: Excho.Core/Repository.fs
namespace Excho.Data

open System.Collections
open System.Collections.Generic
open System.Data.Objects

[<AllowNullLiteral>]
type IRepository =
  abstract member All : obj seq
  abstract member Commit : obj -> unit
  abstract member Delete : obj -> unit

type RepositoryImpl<'t when 't : not struct> = 't IObjectSet

type Repository<'t when 't : not struct>(impl : 't RepositoryImpl) =
  member this.Impl = impl
  abstract member All : 't seq
  default this.All = impl :> seq<_>
  member this.Commit(data) = this.Impl.AddObject(data)
  member this.Delete(data) = this.Impl.DeleteObject(data)
  member this.GetEnumerator() = this.All.GetEnumerator()
  interface IRepository with
    member this.All = this.All :?> seq<_>
    member this.Commit data = 
      let data = data :?> 't
      this.Commit data
    member this.Delete data =
      let data = data :?> 't
      this.Delete data
  interface IEnumerable<'t> with
    member this.GetEnumerator() = this.GetEnumerator()
    member this.GetEnumerator() = this.GetEnumerator() :> IEnumerator